### DISCUSSION TITLE
Lesson 7: EVMError: Revert in testFundUpdatesFundedDataStructure()

### DISCUSSION THREAD

**Discussion by EthanHuntIMF99** [Cyfrin/foundry-full-course-cu#711] [Q&A] [answered]
This is the only test that is failing I am attaching the code here please help me out ;)

![Screenshot (463)](https://github.com/Cyfrin/foundry-full-course-f23/assets/142003515/282ceaa8-9457-44bd-9609-2210b6a58f74)

**Comment by theirrationalone** [comment]
Please make sure you have that `addressToAmountFunded` function into your `FundMe.sol`, after this verification if you still facing this issue then please share `FundMe.sol` also...

**Reply by EthanHuntIMF99** [reply]
If you mean that  getter function then yes I do have that and below is the copy of FundMe.sol  (thxn for reaching out to help in advance):

// SPDX-License-Identifier: MIT
pragma solidity ^0.8.19;

import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
import {PriceConverter} from "./PriceConverter.sol";

error FundMe__NotOwner();

contract FundMe {
    using PriceConverter for uint256;

    mapping(address => uint256) private s_addressToAmountFunded;
    address[] private s_funders;

    // Could we make this constant?  /* hint: no! We should make it immutable! */
    address public /* immutable */ i_owner;
    uint256 public constant MINIMUM_USD = 5 * 10 ** 18;
    AggregatorV3Interface private s_priceFeed;
    
    constructor(address priceFeed) {
        i_owner = msg.sender;
        s_priceFeed = AggregatorV3Interface(priceFeed);
    }

    function fund() public payable {
        require(msg.value.getConversionRate(s_priceFeed ) >= MINIMUM_USD, "You need to spend more ETH!");
        // require(PriceConverter.getConversionRate(msg.value) >= MINIMUM_USD, "You need to spend more ETH!");
        s_addressToAmountFunded[msg.sender] += msg.value;
        s_funders.push(msg.sender);
    }
    
    function getVersion() public view returns (uint256){
        return s_priceFeed.version();
    }
    
    modifier onlyOwner {
        // require(msg.sender == owner);
        if (msg.sender != i_owner) revert FundMe__NotOwner();
        _;
    }
    
    function withdraw() public onlyOwner {
        for (uint256 funderIndex=0; funderIndex < s_funders.length; funderIndex++){
            address funder = s_funders[funderIndex];
            s_addressToAmountFunded[funder] = 0;
        }
        s_funders = new address[](0);
        // // transfer
        // payable(msg.sender).transfer(address(this).balance);
        
        // // send
        // bool sendSuccess = payable(msg.sender).send(address(this).balance);
        // require(sendSuccess, "Send failed");

        // call
        (bool callSuccess, ) = payable(msg.sender).call{value: address(this).balance}("");
        require(callSuccess, "Call failed");
    }
    // Explainer from: https://solidity-by-example.org/fallback/
    // Ether is sent to contract
    //      is msg.data empty?
    //          /   \ 
    //         yes  no
    //         /     \
    //    receive()?  fallback() 
    //     /   \ 
    //   yes   no
    //  /        \
    //receive()  fallback()
    function getAddressToAmountFunded(address FundingAddress) external view returns(uint256){
        return s_addressToAmountFunded[FundingAddress];
    }
    function getFunder(uint56 index) external view returns(address) {
        return s_funders[index];                
    }

    fallback() external payable {
        fund();
    }

    receive() external payable {
        fund();
    }

}

// Concepts we didn't cover yet (will cover in later sections)
// 1. Enum
// 2. Events
// 3. Try / Catch
// 4. Function Selector
// 5. abi.encode / decode
// 6. Hash with keccak256
// 7. Yul / Assembly

**Reply by theirrationalone** [reply]
You got error here at this line: `require(msg.value.getConversionRate(s_priceFeed )`

Please also share `Priceconverter.sol`... 

You might used external visibility modifier with this function `getConversionRate`

**Reply by EthanHuntIMF99** [reply]
Here is the code for PriceConverter.sol:





// SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";

// Why is this a library and not abstract?
// Why not an interface?
library PriceConverter {
    // We could make this public, but then we'd have to deploy it
    function getPrice(AggregatorV3Interface priceFeed) internal view returns (uint256) {
        // Sepolia ETH / USD Address
        // https://docs.chain.link/data-feeds/price-feeds/addresses
        
        
        (, int256 answer, , , ) = priceFeed.latestRoundData();
        // ETH/USD rate in 18 digit
        return uint256(answer * 10000000000);
    }

    // 1000000000
    function getConversionRate(
        uint256 ethAmount,
        AggregatorV3Interface priceFeed
    ) internal view returns (uint256) {

        uint256 ethPrice = getPrice(priceFeed);
        uint256 ethAmountInUsd = (ethPrice * ethAmount) / 1000000000000000000;
        // the actual ETH/USD conversion rate, after adjusting the extra 0s.
        return ethAmountInUsd;
    }
}

**Reply by theirrationalone** [reply]
Your contracts are looking perfect....!

You might assigned some high/low value of `INITIAL_ANSWER` or `DECIMALS` at helperConfig.... Please verify that also , Or share your all scripts... and Please Make sure you have proper mock for AggregatorV3!

**Reply by EthanHuntIMF99** [reply]
Here is the code for HelperConfig.sol


//SPDX-License-Identifier:MIT

pragma solidity^0.8.18;
import {Script} from "forge-std/Script.sol";
import {MockV3Aggregator} from "../test/mocks/MockV3Aggregator.sol";
contract HelperConfig is Script{

    // If we are on a local anvil, we deploy mocks
    // Otherwise,grab the existing address from the live network
    //1. Deploy mocks when we are on a local anvil chain
    //2. Keep track of contract address accross different chains
    //Sepolia Eth/USD
    //Mainnet Eth/USD

    NetworkConfig public activeNetworkConfig;
    
    uint8 public constant DECIMALS=8;
    int256 public constant INITIAL_PRICE=2000e8;

    struct NetworkConfig{
        address priceFeed;
        //ETH/USD price feed address
    }

    constructor (){
        if(block.chainid==11155111){
            activeNetworkConfig = getSepoliaEthConfig();
        }
        else if(block.chainid==1){
            activeNetworkConfig = getMainnetEthConfig();
        }
        else{
            activeNetworkConfig = getOrCreateAnvilEthConfig();
        } 
    }

    function getSepoliaEthConfig() public pure returns (NetworkConfig memory) {
        // price feed address
        NetworkConfig memory seploiaConfig= NetworkConfig({
            priceFeed:0x694AA1769357215DE4FAC081bf1f309aDC325306
        });
        return seploiaConfig;
    }
    function getMainnetEthConfig() public pure returns(NetworkConfig memory){
        //price feed address
        NetworkConfig memory ethConfig= NetworkConfig({
            priceFeed:0x5f4eC3Df9cbd43714FE2740f5E3616155c5b8419
        });
        return ethConfig;
        
    }
    function getOrCreateAnvilEthConfig() public returns(NetworkConfig memory){
        // price feed address
        if (activeNetworkConfig.priceFeed != address(0)){
            return activeNetworkConfig;
        }

        //1.Deploy the mocks
        //2.Return the mock address
        vm.startBroadcast();
        MockV3Aggregator mockPriceFeed = new MockV3Aggregator(
            DECIMALS,
            INITIAL_PRICE
        );
        vm.stopBroadcast();
        NetworkConfig memory anvilConfig = NetworkConfig({
            priceFeed : address(mockPriceFeed)
        });
        return anvilConfig;
    }
}




The code for MockV3Aggregator

// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

/**
 * @title MockV3Aggregator
 * @notice Based on the FluxAggregator contract
 * @notice Use this contract when you need to test
 * other contract's ability to read data from an
 * aggregator contract, but how the aggregator got
 * its answer is unimportant
 */
contract MockV3Aggregator {
    uint256 public constant version = 4;

    uint8 public decimals;
    int256 public latestAnswer;
    uint256 public latestTimestamp;
    uint256 public latestRound;

    mapping(uint256 => int256) public getAnswer;
    mapping(uint256 => uint256) public getTimestamp;
    mapping(uint256 => uint256) private getStartedAt;

    constructor(uint8 _decimals, int256 _initialAnswer) {
        decimals = _decimals;
        updateAnswer(_initialAnswer);
    }

    function updateAnswer(int256 _answer) public {
        latestAnswer = _answer;
        latestTimestamp = block.timestamp;
        latestRound++;
        getAnswer[latestRound] = _answer;
        getTimestamp[latestRound] = block.timestamp;
        getStartedAt[latestRound] = block.timestamp;
    }

    function updateRoundData(uint80 _roundId, int256 _answer, uint256 _timestamp, uint256 _startedAt) public {
        latestRound = _roundId;
        latestAnswer = _answer;
        latestTimestamp = _timestamp;
        getAnswer[latestRound] = _answer;
        getTimestamp[latestRound] = _timestamp;
        getStartedAt[latestRound] = _startedAt;
    }

    function getRoundData(uint80 _roundId)
        external
        view
        returns (uint80 roundId, int256 answer, uint256 startedAt, uint256 updatedAt, uint80 answeredInRound)
    {
        return (_roundId, getAnswer[_roundId], getStartedAt[_roundId], getTimestamp[_roundId], _roundId);
    }

    function latestRoundData()
        external
        view
        returns (uint80 roundId, int256 answer, uint256 startedAt, uint256 updatedAt, uint80 answeredInRound)
    {
        return (
            uint80(latestRound),
            getAnswer[latestRound],
            getStartedAt[latestRound],
            getTimestamp[latestRound],
            uint80(latestRound)
        );
    }

    function description() external pure returns (string memory) {
        return "v0.6/test/mock/MockV3Aggregator.sol";
    }
}

Code for DeployFundMe:

//SPDX-License-Identifier:MIT
pragma solidity^0.8.18;

import {Script} from "forge-std/Script.sol";
import {FundMe} from "../src/FundMe.sol";

contract DeployFundMe is Script{
    function run() external returns(FundMe){
        vm.startBroadcast();
        FundMe fundMe = new FundMe(0x694AA1769357215DE4FAC081bf1f309aDC325306);
        vm.stopBroadcast();
        return fundMe;
    }
}


These are all the remaining codes please help me out ;)

**Comment by EthanHuntIMF99** [comment, marked as answer]
Thanks for your help @theirrationalone  I found out where I was making mistake in the DeployFundMe.s.sol I forgot to import HelperConfig.sol and the code regarding activeNetworkConfig . Thanks for the help tho really appreciate that.
